File: 1. Creational Design Patterns/Abstract Factory/Typescript/AbstractFactoryTemplate.ts
// Abstract Factory Pattern – TypeScript Template

// Abstract Products
interface ProductA {}
interface ProductB {}

// Concrete Products
class ConcreteProductA1 implements ProductA {}
class ConcreteProductB1 implements ProductB {}

// Abstract Factory
interface AbstractFactory {
  createProductA(): ProductA;
  createProductB(): ProductB;
}

// Concrete Factory 1
class ConcreteFactory1 implements AbstractFactory {
  createProductA() { return new ConcreteProductA1(); }
  createProductB() { return new ConcreteProductB1(); }
}

// Client Skeleton
const factory: AbstractFactory = new ConcreteFactory1();
// const productA = factory.createProductA();

File: 1. Creational Design Patterns/Abstract Factory/Go/AbstractFactoryTemplate.go
// Abstract Factory Pattern – Go Skeleton / Template
// Replace products & factories according to your domain logic.
package main

// TODO: Define abstract products

type ProductA interface{}
type ProductB interface{}

// TODO: Define concrete products

type ConcreteProductA1 struct{}
type ConcreteProductB1 struct{}

// TODO: Define abstract factory

type AbstractFactory interface {
	CreateProductA() ProductA
	CreateProductB() ProductB
}

// TODO: Implement concrete factory

type ConcreteFactory1 struct{}

func (ConcreteFactory1) CreateProductA() ProductA { return ConcreteProductA1{} }
func (ConcreteFactory1) CreateProductB() ProductB { return ConcreteProductB1{} }

func main() {
	// TODO: Client code
}
